### PR TITLE
KOGITO-1747: Cannot use list data type inside structure

### PR DESCRIPTION
**Pull request by yesamer**
@danielezonca @dupliaka Can you please review and test it?

Main fixes:
- isCollection fields are fixed;
- nestedField are now supported for `ItemComponent` as well
- If a InputData or a Decision node has empty `typeRef`, a `Any` `DmnType` is assigned.

https://issues.redhat.com/browse/KOGITO-1747`

### CODE REVIEW

**Comment by dupliaka** [comment]
Caused by https://issues.redhat.com/browse/DROOLS-5250.

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-kogito-client/src/main/java/org/drools/workbench/screens/scenariosimulation/kogito/client/dmn/AbstractKogitoDMNService.java [hunk 1]
```diff
@@ -90,10 +89,21 @@ protected ClientDMNType getDMNTypeFromMaps(final Map<String, ClientDMNType> dmnT
         return dmnTypesMap.get(typeRef);
     }
 
+    /**
+     * This method retrieves all DMN Types are composed by
+     * - Primitive FEEL types, listed inside <code>BuiltInType</code> enum
+     * - User defined Types, which are available inside inside <code>jsitItemDefinitions</code> list
+     * @param jsitItemDefinitions
```
Comment (dupliaka) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-kogito-client/src/main/java/org/drools/workbench/screens/scenariosimulation/kogito/client/dmn/AbstractKogitoDMNService.java:96]: Shouldn't it be jsItItemDefinition?

**Comment by dupliaka** [comment]
In Kogito If you select a fact with context type then it is not selects in test tools see
image

**Comment by dupliaka** [comment]
ContextType.value could be duplicated 
image

**Comment by dupliaka** [comment]
3. In Kogito Nor Nested or context type could be added as expression 
image

**Comment by dupliaka** [comment]
4. Scesim based on DMN with collections is not opening
https://issues.redhat.com/secure/attachment/12471374/CollectionsDoesNotWork.webm

**Comment by dupliaka** [comment]
5. Test tools data objects sorted differently in BC and Kogito: 
BC
image
Kogito
image

**Comment by dupliaka** [comment]
The issues I found for Kogito and BC
- [ ] 1. Test tools selector
- [ ] 2. Duplicates in table
- [ ] 3. Expressions for nested and context
- [ ] 4. Collections for nested
- [ ] 5. Reorder test tools
- [ ] 6. [Fixes RHDM-1153](https://issues.redhat.com/browse/RHDM-1153?focusedCommentId=14055844&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14055844)

**Comment by yesamer** [comment]
1 and 2. It reproduces in BC as well. Please open a new ticket for it. (eg. Issues with DMN context type);
3. Not a bug: Expressions can be set only for first-level Instances;
4. _... checking ...._
5. Please open a new ticket for it (low priority)
6. I'll manage that in another PR

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-kogito-client/src/main/java/org/drools/workbench/screens/scenariosimulation/kogito/client/dmn/AbstractKogitoDMNService.java (hunk 1 above)
Comment (yesamer) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-kogito-client/src/main/java/org/drools/workbench/screens/scenariosimulation/kogito/client/dmn/AbstractKogitoDMNService.java:96]: @dupliaka `jsitItemDefinitions` is the list name, I think it's correct.

**Comment by yesamer** [comment]
@dupliaka Improved nested fields managing and collections fields as well. Can you please have a look again?

**Comment by dupliaka** [comment]
7. Nested context list impossible to edit 
image

**Comment by yesamer** [comment]
@dupliaka Thank you, I fixed the last case you reported as well. I performed some manual test this morning, I think most cases should be correctly handled. Can you please restest it? In the meanwhile, I'll work on Unit Tests.

**Comment by dupliaka** [comment]
> @dupliaka Thank you, I fixed the last case you reported as well. I performed some manual test this morning, I think most cases should be correctly handled. Can you please retest it? In the meanwhile, I'll work on Unit Tests.

I tested that manually and get the same result the collection editor did not appears for lost of context. Also the fact is not getting highlighted in test tool panel. 
https://issues.redhat.com/secure/attachment/12471793/contextCollection.webm

**Comment by yesamer** [comment]
@dupliaka I suggest to don't consider issue related to `context` type for the moment, If I remember correctly you opened some specific tickets for `content` management. Please update these tickets reporting this case, and for this PR please don't consider `context` type. 
Thanks!

**Comment by dupliaka** [comment]
> @dupliaka I suggest to don't consider issue related to `context` type for the moment, If I remember correctly you opened some specific tickets for `content` management. Please update these tickets reporting this case (or open a new one), and for this PR please don't consider `context` type.
> Thanks!

Ok, thanks. Then manual check is finished.

**Comment by yesamer** [comment]
jenkins retest this please

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-kogito-client/src/main/java/org/drools/workbench/screens/scenariosimulation/kogito/client/dmn/ClientDMNType.java [hunk 2]
```diff
@@ -37,12 +38,30 @@ public ClientDMNType(String namespace, String name, String id, boolean isCollect
...
+
     public ClientDMNType(String namespace, String name, String id, boolean isCollection, boolean isComposite, Map<String, ClientDMNType> fields, BuiltInType feelType) {
         this(namespace, name, id, isCollection, feelType);
         this.fields = fields;
         this.composite = isComposite;
     }
 
+    /**
+     * It *copies* the current ClientDMNType, setting it as Collection, with isCollection = true.
+     * Please note, fields parameter reference should be the same for both object, the original one
```
[the remark this answers is not in the document]
Comment (yesamer) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-kogito-client/src/main/java/org/drools/workbench/screens/scenariosimulation/kogito/client/dmn/ClientDMNType.java:59]: @danielezonca Thank you, done.

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-kogito-client/src/main/java/org/drools/workbench/screens/scenariosimulation/kogito/client/dmn/AbstractKogitoDMNService.java [hunk 3]
```diff
@@ -136,94 +154,132 @@ ClientDMNType getOrCreateDMNType(final Map<String, JSITItemDefinition> allDefini
...
+                itemDefinitionDMNType.setCollection(superDmnType.isCollection() || itemDefinitionDMNType.isCollection());
+                itemDefinitionDMNType.setIsComposite(superDmnType.isComposite() || itemDefinitionDMNType.isComposite());
+                itemDefinitionDMNType.setFeelType(superDmnType.getFeelType() != null ? superDmnType.getFeelType() : null);
             } else {
                 throw new IllegalStateException(
                         "Item: " + itemDefinition.getName() + " refers to typeRef: " + itemDefinition.getTypeRef()
                                 + " which can't be found.");
             }
+        } else {
+            populateItemDefinitionFields(allDefinitions,
```
[the remark this answers is not in the document]
Comment (yesamer) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-kogito-client/src/main/java/org/drools/workbench/screens/scenariosimulation/kogito/client/dmn/AbstractKogitoDMNService.java]: @danielezonca done.

**Comment by yesamer** [comment]
@kiegroup/gatekeepers can you please merge it?

**Pull request merged**